File: dashboard/src/lib/api/approvals.ts
import { request } from './client';
import type { PendingCall } from './types';

export interface PendingRun {
	run_id: string;
	agent_name: string;
	role_path: string | null;
	created_at: string;
	originating_prompt: string | null;
	calls: PendingCall[];
}

export interface PendingListResponse {
	runs: PendingRun[];
	count: number;
}

export interface PendingCountResponse {
	count: number;
}

export interface ApprovalsResolveRequest {
	decisions: Record<string, boolean>;
	resolved_by?: string | null;
}

export interface ApprovalsResolveResponse {
	run_id: string;
	status: 'done' | 'paused';
	success: boolean;
	output: string;
	error: string | null;
	tokens_in: number;
	tokens_out: number;
	total_tokens: number;
	duration_ms: number;
	message_history: string | null;
	pending_approvals: PendingCall[];
}

export function listPending(limit = 200): Promise<PendingListResponse> {
	return request(`/api/approvals/pending?limit=${limit}`);
}

export function countPending(): Promise<PendingCountResponse> {
	return request('/api/approvals/pending?count_only=1');
}

export function getPendingRun(runId: string): Promise<PendingRun> {
	return request(`/api/approvals/${encodeURIComponent(runId)}`);
}

export function resolveRun(
	runId: string,
	req: ApprovalsResolveRequest
): Promise<ApprovalsResolveResponse> {
	return request(`/api/approvals/${encodeURIComponent(runId)}`, {
		method: 'POST',
		body: JSON.stringify(req)
	});
}

File: initrunner/dashboard/schemas/approvals.py
"""Pydantic response schemas for the approvals router."""

from __future__ import annotations

from typing import Any

from pydantic import BaseModel


class PendingCallResponse(BaseModel):
    """One tool call awaiting human approval."""

    tool_call_id: str
    tool_name: str
    arguments: dict[str, Any]


class PendingRunResponse(BaseModel):
    """One paused run with every tool call awaiting approval on it.

    The dashboard queue groups by run because
    ``services.execution.resume_run_sync`` requires decisions for every
    ``tool_call_id`` in a run — partial batches are rejected.
    """

    run_id: str
    agent_name: str
    role_path: str | None
    created_at: str
    originating_prompt: str | None
    calls: list[PendingCallResponse]


class PendingListResponse(BaseModel):
    """Queue payload: grouped pending runs + total count."""

    runs: list[PendingRunResponse]
    count: int


class PendingCountResponse(BaseModel):
    """Lightweight nav-badge payload."""

    count: int


class ApprovalsResolveRequest(BaseModel):
    """Body of ``POST /api/approvals/{run_id}``: one bool per tool_call_id."""

    decisions: dict[str, bool]
    resolved_by: str | None = None


class ApprovalsResolveResponse(BaseModel):
    """Result of a resume.

    - ``status == "done"``: the run completed; ``output`` carries the result.
    - ``status == "paused"``: the model queued more approval-required calls;
      ``pending_approvals`` lists them and the caller should loop.
    """

    run_id: str
    status: str  # "done" | "paused"
    success: bool
    output: str
    error: str | None
    tokens_in: int
    tokens_out: int
    total_tokens: int
    duration_ms: int
    message_history: str | None
    pending_approvals: list[PendingCallResponse]
